File: HW-05/PolygonVect.cpp
/*
 * Gokhan Has - 161044067
*/

#include "PolygonVect.h"

namespace HasShapes{

	PolygonVect::PolygonVect() noexcept {/* Body Initialize Empty */}
	
	PolygonVect::PolygonVect(const Rectangle& otherRectangle)noexcept{

		areaPolygon = otherRectangle.area();
		perimeterPolygon = otherRectangle.perimeter();
		
		Polygon::Point2D temp;
		
		temp.setCoordinate_X(otherRectangle.get_x());
		temp.setCoordinate_Y(otherRectangle.get_y());
		vectorPoint2D.push_back(temp);

		temp.setCoordinate_X(otherRectangle.get_x() + otherRectangle.getWidth());
		temp.setCoordinate_Y(otherRectangle.get_y());
		vectorPoint2D.push_back(temp);

		temp.setCoordinate_X(otherRectangle.get_x() + otherRectangle.getWidth());
		temp.setCoordinate_Y(otherRectangle.get_x() + otherRectangle.getHeight());
		vectorPoint2D.push_back(temp);

		temp.setCoordinate_X(otherRectangle.get_x());
		temp.setCoordinate_Y(otherRectangle.get_y() + otherRectangle.getHeight());
		vectorPoint2D.push_back(temp);
	}
	
	PolygonVect::PolygonVect(const Triangle& otherTriangle)noexcept{

		areaPolygon = otherTriangle.area();
		perimeterPolygon = otherTriangle.perimeter();

		Polygon::Point2D temp;

		if(otherTriangle.getRotateControl() == 0){ // Normal triangle's coordinate 
			
			temp.setCoordinate_X(otherTriangle.get_x());
			temp.setCoordinate_Y(otherTriangle.get_y());
			vectorPoint2D.push_back(temp);

			temp.setCoordinate_X(otherTriangle.get_x() - otherTriangle.getEdge()/2);
			temp.setCoordinate_Y(otherTriangle.get_y() + otherTriangle.getEdge()/2*sqrt(3));
			vectorPoint2D.push_back(temp);

			temp.setCoordinate_X(otherTriangle.get_x() + otherTriangle.getEdge()/2);
			temp.setCoordinate_Y(otherTriangle.get_y() + otherTriangle.getEdge()/2*sqrt(3));
			vectorPoint2D.push_back(temp);
		}

		else{ // Inverse triangle's coordinate

			temp.setCoordinate_X(otherTriangle.get_x());
			temp.setCoordinate_Y(otherTriangle.get_y());
			vectorPoint2D.push_back(temp);

			temp.setCoordinate_X(otherTriangle.get_x() + otherTriangle.getEdge());
			temp.setCoordinate_Y(otherTriangle.get_y());
			vectorPoint2D.push_back(temp);

			temp.setCoordinate_X(otherTriangle.get_x() + otherTriangle.getEdge()/2);
			temp.setCoordinate_Y(otherTriangle.get_y() + otherTriangle.getEdge()/2*sqrt(3));
			vectorPoint2D.push_back(temp);			
		}
	}
	
	PolygonVect::PolygonVect(const Circle& otherCircle)noexcept{

		areaPolygon = otherCircle.area();
		perimeterPolygon = otherCircle.perimeter();

		Polygon::Point2D temp;
		int size = 100;
		
		double point_X = otherCircle.get_x();
		double point_Y = otherCircle.get_y() - otherCircle.getRadius();
		
		temp.setCoordinate_X(point_X);
		temp.setCoordinate_Y(point_Y);
		vectorPoint2D.push_back(temp);	

   		for(int i=1; i<size; ++i){

	    	temp.setCoordinate_X( otherCircle.get_x() + (otherCircle.getRadius()*sin(static_cast<double>(i) * PI/50.0)));
	    	temp.setCoordinate_Y( otherCircle.get_y() + ((-1)*otherCircle.getRadius()*cos(static_cast<double>(i) * PI/50.0)));
	    	vectorPoint2D.push_back(temp);
	    }
	}
	
	PolygonVect::PolygonVect(const PolygonVect& copy)noexcept{

		for(int i=0;i<copy.vectorPoint2D.size();++i)
			vectorPoint2D[i] = copy.vectorPoint2D[i];
		controlShapeColor = copy.controlShapeColor;
	}
	
	PolygonVect& PolygonVect::operator =(const PolygonVect& other)noexcept{

		if(this != &other){

			vectorPoint2D.clear();
			
			for(int i=0;i<other.vectorPoint2D.size();++i)
				vectorPoint2D[i] = other.vectorPoint2D[i];
			controlShapeColor = other.controlShapeColor;
		}
		return *this;
	}
	
	PolygonVect::~PolygonVect() noexcept {/* Body Initialize Empty */}
	
	void PolygonVect::newPrintSVG(ostream& outs) noexcept{

		outs << "\n<polygon points = ";
		outs << "\"";
		
		for(int i=0;i<this->vectorPoint2D.size();++i)
			outs << this->vectorPoint2D[i].getX() << "," << this->vectorPoint2D[i].getY() << " ";
		
		if(this->controlShapeColor == 0) // mainShape color
			outs << "\" style=\"fill:red;\"/>";
		else if(this->controlShapeColor == 1)
			outs << "\" style=\"fill:purple;\"/>"; // smallShape color
		else
			outs << "\" style=\"fill:green;\"/>";
	}
}

File: HW-07/161044067_CSE241_HW7_project/src/PolygonVect.java
/*
 * Gokhan Has - 161044067
 */


import java.awt.Color;
import java.awt.Graphics;
import java.util.ArrayList;

/**
 * This is the class PolygonVect extends hw07.Polygon class.
 * Keeps the ArrayList for coordinates(Point2D).
 * @author Gokhan Has
 *
 */
public class PolygonVect extends Polygon {

	private ArrayList<Point2D> arrayList = new ArrayList<Point2D> ();
	
	public PolygonVect() {
		arrayList = null;
	}
	/**
	 * Convert Rectangle to PolygonVect.
	 * @param rectangle which converted PolygonVect.
	 * @throws Exception if the x or y is negative value.
	 */
	public PolygonVect(Rectangle rectangle) throws Exception {
		
		areaPolygon = rectangle.area();
		perimeterPolygon = rectangle.perimeter();
		
		Point2D temp = new Point2D();
		
		temp.setPos_x(rectangle.getPos_x());
		temp.setPos_y(rectangle.getPos_y());
		arrayList.add(temp);
		
		temp = new Point2D();
		temp.setPos_x(rectangle.getPos_x() + rectangle.get_Width());
		temp.setPos_y(rectangle.getPos_y());
		arrayList.add(temp);
	
		temp = new Point2D();
		temp.setPos_x(rectangle.getPos_x() + rectangle.get_Width());
		temp.setPos_y(rectangle.getPos_x() + rectangle.get_Height());
		arrayList.add(temp);

		temp = new Point2D();
		temp.setPos_x(rectangle.getPos_x());
		temp.setPos_y(rectangle.getPos_y() + rectangle.get_Height());
		arrayList.add(temp);
		
	}
	
	/**
	 * Convert Circle to PolygonVect.
	 * @param circle which converted PolygonVect.
	 * @throws Exception if the x or y is negative value.
	 */
	public PolygonVect(Circle circle) throws Exception {
		
		areaPolygon = circle.area();
		perimeterPolygon = circle.perimeter();
		
		int size = 100;
		
		double point_X = circle.getPos_x();
		double point_Y = circle.getPos_y() - circle.getRadius();
		
		Point2D temp = new Point2D();
		temp.setPos_x(point_X);
		temp.setPos_y(point_Y);
		arrayList.add(temp);

   		for(int i=1; i<size; ++i){
   			
   			temp = new Point2D();
	    	temp.setPos_x( circle.getPos_x() + (circle.getRadius()*Math.sin((i) * Math.PI/50.0)));
	    	temp.setPos_y( circle.getPos_y() + ((-1)*circle.getRadius()*Math.cos((i) * Math.PI/50.0)));
	    	arrayList.add(temp);
	    }
	}
	/**
	 * Convert Triangle to PolygonVect.
	 * @param triangle which converted PolygonVect.
	 * @throws Exception if the x or y is negative value.
	 */
	public PolygonVect(Triangle triangle) throws Exception {
		
		areaPolygon = triangle.area();
		perimeterPolygon = triangle.perimeter();
		
		Point2D temp = new Point2D();
		
		if(triangle.getRotateControl() == 0){ // Normal triangle's coordinate 
			
			temp.setPos_x(triangle.getPos_x());
			temp.setPos_y(triangle.getPos_y());
			arrayList.add(temp);

			temp = new Point2D();
			temp.setPos_x(triangle.getPos_x() - triangle.getEdge()/2);
			temp.setPos_y(triangle.getPos_y() + triangle.getEdge()/2*Math.sqrt(3));
			arrayList.add(temp);
			
			temp = new Point2D();
			temp.setPos_x(triangle.getPos_x() + triangle.getEdge()/2);
			temp.setPos_y(triangle.getPos_y() + triangle.getEdge()/2*Math.sqrt(3));
			arrayList.add(temp);
		}

		else{ // Inverse triangle's coordinate

			temp.setPos_x(triangle.getPos_x());
			temp.setPos_y(triangle.getPos_y());
			arrayList.add(temp);

			temp = new Point2D();
			temp.setPos_x(triangle.getPos_x() + triangle.getEdge());
			temp.setPos_y(triangle.getPos_y());
			arrayList.add(temp);

			temp = new Point2D();
			temp.setPos_x(triangle.getPos_x() + triangle.getEdge()/2);
			temp.setPos_y(triangle.getPos_y() + triangle.getEdge()/2*Math.sqrt(3));
			arrayList.add(temp);			
		}	
	}

	public ArrayList<Point2D> getArrayList() {
		return arrayList;
	}

	public void setArrayList(ArrayList<Point2D> arrayList) {
		this.arrayList = arrayList;
	}
	
	@Override
	public double area() {
		return areaPolygon;
	}

	@Override
	public double perimeter() {
		return perimeterPolygon;
	}

	@Override
	public void increment() throws Exception {
		for(int i=0;i<arrayList.size();++i) {
			arrayList.get(i).setPos_x(arrayList.get(i).getPos_x()+1);
			arrayList.get(i).setPos_y(arrayList.get(i).getPos_y()+1);
		}
	}

	@Override
	public void decrement() throws Exception {
		for(int i=0;i<arrayList.size();++i) {
			arrayList.get(i).setPos_x(arrayList.get(i).getPos_x()-1);
			arrayList.get(i).setPos_y(arrayList.get(i).getPos_y()-1);
		}
	}

	@Override
	public void draw(Graphics g) {
		paintComponent(g);
	}

	public void paintComponent(Graphics g) {
		super.paintComponent(g);
		int [] xPoints = new int[arrayList.size()];
		int [] yPoints = new int[arrayList.size()];
		
		for(int i=0;i<arrayList.size();++i) {
			xPoints[i] = (int) arrayList.get(i).getPos_x();
			yPoints[i] = (int) arrayList.get(i).getPos_y();
		}
		g.setColor(Color.BLUE);
		g.drawPolygon(xPoints, yPoints,arrayList.size());
	}

	@Override
	public int compareTo(Shape other) {
		if(this.area() == other.area())
			return 0;
		else if(this.area() > other.area())
			return 1;
		else
			return -1;
	}
}
